Src/Plugins/Security/SEC101_055.CmdkeyValidator.cs

```diff
@@ -0,0 +1,116 @@
+using System.Linq;
+
+using Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security.Utilities;
+using Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk;
+using Microsoft.RE2.Managed;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security
+{
+    public class CmdkeyValidator : ValidatorBase
+    {
+        internal static CmdkeyValidator Instance;
+        internal static IRegex RegexEngine;
+
+        private const string UserRegex = @"\/user:\s*([^\[""'\s`%]{2,})";
+
+        static CmdkeyValidator()
+        {
+            Instance = new CmdkeyValidator();
+            RegexEngine = RE2Regex.Instance;
+        }
+
+        public static string IsValidStatic(ref string matchedPattern,
+                                           ref Dictionary<string, string> groups,
+                                           ref string failureLevel,
+                                           ref string fingerprint,
+                                           ref string message)
+        {
+            return IsValidStatic(Instance,
+                                 ref matchedPattern,
+                                 ref groups,
+                                 ref failureLevel,
+                                 ref fingerprint,
+                                 ref message);
+        }
+
+        protected override string IsValidStaticHelper(ref string matchedPattern,
+                                                      ref Dictionary<string, string> groups,
+                                                      ref string failureLevel,
+                                                      ref string fingerprintText,
+                                                      ref string message)
+        {
+            if (!groups.TryGetNonEmptyValue("password", out string password))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            if (IsInvalidPassword(password))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            string account = ParseExpression(RegexEngine, matchedPattern, UserRegex);
+            account = account?.Replace("/user:", string.Empty).Trim();
+
+            fingerprintText = new Fingerprint()
+            {
+                Account = account,
+                Password = password,
+            }.ToString();
+
+            // Manual use of the regex against CodeAsData produces highly confident results,
+            // so return AuthorizedWarning rather than Unknown.
+            return nameof(ValidationState.AuthorizedWarning);
+        }
+
+        private static bool IsInvalidPassword(string potentialPassword)
+        {
+            if (string.IsNullOrWhiteSpace(potentialPassword))
+            {
+                return true;
+            }
+
+            if (potentialPassword.Length < 2)
+            {
+                return true;
+            }
+
+            if (potentialPassword.First() == '$')
```

Comment:
>potentialPassword [](start = 16, length = 17)

Just index into the string for this, potentialPassword[0]. But really, do we want this? Couldn't this character exist in a BAT or CMD file? Not for PS of course.

Reply:
Brainfart, that's much simpler.

It's hard to say what will be valid characters in all contexts, I suspect that in some contexts $ is allowed.  And the regex for this validator does indeed allow it...except for the very first character.  So yes, we may miss some passwords, but only those passwords which put $ in the first position.  In the data I viewed, I saw *many* more potential false positives where a powershell variable was being used.  So I think this is a worthwhile tradeoff.  Let me know if you think differently.

---
In reply to: [598012251](https://github.com/microsoft/sarif-pattern-matcher/pull/317#discussion_r598012251) [](ancestors = 598012251)